module/zfs/ddt.c

```diff
@@ -816,8 +816,8 @@ ddt_entry_compare(const void *x1, const void *x2)
 {
 	const ddt_entry_t *dde1 = x1;
 	const ddt_entry_t *dde2 = x2;
-	const uint64_t *u1 = (const uint64_t *)&dde1->dde_key;
-	const uint64_t *u2 = (const uint64_t *)&dde2->dde_key;
+	const uint64_t **u1 = (const uint64_t **)&dde1->dde_key;
+	const uint64_t **u2 = (const uint64_t **)&dde2->dde_key;
```

Comment:
This isn't going to work for 32bit builds. Some more conventional type punning should be used. Here's an example:
```diff
diff --git a/module/zfs/ddt.c b/module/zfs/ddt.c
index 2e760f2..054f71c 100644
--- a/module/zfs/ddt.c
+++ b/module/zfs/ddt.c
@@ -814,16 +814,21 @@ ddt_prefetch(spa_t *spa, const blkptr_t *bp)
 int
 ddt_entry_compare(const void *x1, const void *x2)
 {
+       typedef struct dde_key_cmp {
+               uint64_t dkc_key_u64[DDT_KEY_WORDS];
+       } dde_key_cmp_t;
+
        const ddt_entry_t *dde1 = x1;
        const ddt_entry_t *dde2 = x2;
-       const uint64_t *u1 = (const uint64_t *)&dde1->dde_key;
-       const uint64_t *u2 = (const uint64_t *)&dde2->dde_key;
+       const dde_key_cmp_t *u1 = (dde_key_cmp_t *) &dde1->dde_key;
+       const dde_key_cmp_t *u2 = (dde_key_cmp_t *) &dde2->dde_key;
+
        int i;

        for (i = 0; i < DDT_KEY_WORDS; i++) {
-               if (u1[i] < u2[i])
+               if (u1->dkc_key_u64[i] < u2->dkc_key_u64[i])
                        return (-1);
-               if (u1[i] > u2[i])
+               if (u1->dkc_key_u64[i] > u2->dkc_key_u64[i])
                        return (1);
        }

```